src/script/api/script_group.cpp

```diff
@@ -110,20 +117,27 @@
 
 /* static */ SQInteger ScriptGroup::GetNumEngines(GroupID group_id, EngineID engine_id)
 {
-	EnforceCompanyModeValid(-1);
-	if (!IsValidGroup(group_id) && group_id != GROUP_DEFAULT && group_id != GROUP_ALL) return -1;
+	EnforceDeityOrCompanyModeValid(-1);
```

Comment:
What is the rationale for all this complicated logic, over just requiring company mode for all ScriptGroup calls?
Why is it needed that a GS can iterate over almost, but not all, groups?

Especially because to reach all functionality of this function, I still need to switch to company mode. I really do not see a reason for a GS to iterate (almost) all the groups, I actually do not see much reason for a GS to access groups but I can agree that if adding it takes only five characters there is little harm. However, making the API complicated for something that will barely get used seems overkill.

Reply:
The rationale, if it is querying information without modifying anything in the company, then it should be done without being in the company. Too bad that for this and 2 other functions to retrieve information for GROUP_DEFAULT and GROUP_ALL, a specific company must be provided. The only way for that is to EnforceCompanyModeValid. The other groups are company owned, which makes it possible to query this information without being in the company.

Reply:
You say that if you don't modify the company you should not need to be in the company, but then still require the GS to be in the company to get the complete picture.

And to be advocate's devil: why shouldn't this function, in deity mode, not return the total number of engines of all companies in the default or all group? Similarly, shouldn't ScriptVehicleList_DefaultGroup contain all vehicles that are not in a group if you're deity? After all, in deity mode basically all the other lists are the combined result of all companies. 

That way you do not need to construct complicated preconditions, or add new precondition check functions. And the GS does not need to iterate through all the companies if it wants the combined information.